File: app/src/main/java/org/blackcandy/android/bridge/SearchComponent.kt
package org.blackcandy.android.bridge

import androidx.fragment.app.Fragment
import com.google.android.material.search.SearchBar
import com.google.android.material.search.SearchView
import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.Serializable
import org.blackcandy.android.R

class SearchComponent(
    name: String,
    private val delegate: BridgeDelegate<HotwireDestination>,
) : BridgeComponent<HotwireDestination>(name, delegate) {
    private val fragment: Fragment
        get() = delegate.destination.fragment

    private val searchBar: SearchBar?
        get() = fragment.view?.findViewById(R.id.search_bar)

    private val searchView: SearchView?
        get() = fragment.view?.findViewById(R.id.search_view)

    override fun onReceive(message: Message) {
        when (message.event) {
            "connect" -> handleConnectEvent(message)
        }
    }

    private fun handleConnectEvent(message: Message) {
        searchView?.setupWithSearchBar(searchBar)
        searchView?.editText?.setOnEditorActionListener { _, _, _ ->
            val searchText = searchView?.text.toString()

            if (searchText.isEmpty()) return@setOnEditorActionListener false

            searchView?.hide()
            searchBar?.setText(searchText)
            replyTo("connect", SearchData(query = searchText))

            true
        }
    }

    @Serializable
    data class SearchData(
        val query: String,
    )
}

File: iosApp/iosApp/Bridge/SearchComponent.swift
import HotwireNative
import UIKit

class SearchComponent: BridgeComponent {
    private var viewController: UIViewController? {
        delegate?.destination as? UIViewController
    }

    private lazy var searchBarDelegator = SearchBarDelegator(component: self)

    override class var name: String { "search" }

    override func onReceive(message: Message) {
        switch message.event {
        case "connect":
            handleConnectEvent(message)
        default:
            break
        }
    }

    private func updateSearchResults(with query: String) {
        let data = SearchData(query: query)
        reply(to: "connect", with: data)
    }

    private func handleConnectEvent(_ message: Message) {
        guard let viewController else { return }

        let searchController = UISearchController(searchResultsController: nil)

        searchController.searchBar.delegate = searchBarDelegator

        viewController.navigationItem.searchController = searchController
        viewController.navigationItem.hidesSearchBarWhenScrolling = false
    }

}

extension SearchComponent {
    struct SearchData: Encodable {
        let query: String
    }

    private class SearchBarDelegator: NSObject, UISearchBarDelegate {
        private weak var component: SearchComponent?

        init(component: SearchComponent) {
            self.component = component
        }

        func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
            guard let query = searchBar.searchTextField.text, !query.isEmpty else { return }
            component?.updateSearchResults(with: query)
        }
    }
}
